qiling/loader/pe_uefi.py

```diff
@@ -153,188 +143,156 @@ def map_and_load(self, path, execute_now=False):
-        system_table_heap_ptr += ctypes.sizeof(EFI_DXE_SERVICES)
-        system_table_heap_ptr, dxe_services = install_EFI_DXE_SERVICES(self.ql, system_table_heap_ptr)
-        efi_configuration_table = EFI_CONFIGURATION_TABLE()
-        efi_configuration_table.VendorGuid.Data1 = int(self.ql.os.profile.get("DXE_SERVICE_TABLE", "data1"), 16)
-        efi_configuration_table.VendorGuid.Data2 = int(self.ql.os.profile.get("DXE_SERVICE_TABLE", "data2"), 16)
-        efi_configuration_table.VendorGuid.Data3 = int(self.ql.os.profile.get("DXE_SERVICE_TABLE", "data3"), 16)
-        
-        data4 = ast.literal_eval(self.ql.os.profile.get("DXE_SERVICE_TABLE", "data4"))
-        datalist = 0
-        for data4_list in data4:
-            efi_configuration_table.VendorGuid.Data4[datalist] = data4_list
-            datalist += 1  
-       
-        efi_configuration_table.VendorTable = self.dxe_services_ptr
-        self.ql.mem.write(self.efi_configuration_table_ptr + ctypes.sizeof(EFI_CONFIGURATION_TABLE), convert_struct_to_bytes(efi_configuration_table))
-        self.efi_configuration_table.append(self.ql.os.profile.get("DXE_SERVICE_TABLE", "guid"))
-        
-
-        self.ql.mem.write(self.runtime_services_ptr, convert_struct_to_bytes(self.runtime_services))
-        self.ql.mem.write(self.boot_services_ptr, convert_struct_to_bytes(boot_services))
-        self.ql.mem.write(self.system_table_ptr, convert_struct_to_bytes(system_table))
-        self.ql.mem.write(self.mm_system_table_ptr, convert_struct_to_bytes(efi_mm_system_table))
-        self.ql.mem.write(self.smm_base2_protocol_ptr, convert_struct_to_bytes(smm_base2_protocol))
-        self.ql.mem.write(self.mm_access_protocol_ptr, convert_struct_to_bytes(mm_access_protocol))
-        self.ql.mem.write(self.smm_sw_dispatch2_protocol_ptr, convert_struct_to_bytes(smm_sw_dispatch2_protocol))
-        self.ql.mem.write(self.dxe_services_ptr, convert_struct_to_bytes(dxe_services))
+        self.entry_point = 0
+        self.load_address = 0
+        self.next_image_base = int(os_profile["image_address"], 0)
 
         for dependency in self.ql.argv:
             if not self.map_and_load(dependency):
                 raise QlErrorFileType("Can't map dependency")
 
-        logging.info("[+] Done with loading %s" % self.ql.path)
+        logging.info(f"[+] Done with loading {self.ql.path}")
 
-        #return address
-        self.end_of_execution_ptr = system_table_heap_ptr
+        # hack: reuse first byte of ST to set a trap
+        self.end_of_execution_ptr = gST
         self.ql.mem.write(self.end_of_execution_ptr, b'\xcc')
-        system_table_heap_ptr += pointer_size
         self.ql.hook_address(hook_EndOfExecution, self.end_of_execution_ptr)
 
         self.execute_next_module()
 
     def restore_runtime_services(self):
-        self.ql.mem.write(self.runtime_services_ptr, convert_struct_to_bytes(self.runtime_services))
-
+        pass # not sure why do we need to restore RT
```

Comment:
When running multiple modules, some of them overwrite some runtime_services functions. Then we lose out hooks.